File: src/linkedlist/Solution2.java
package linkedlist;

// 2. 两数相加
public class Solution2 {

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {

        int carry = 0;
        ListNode pre = new ListNode(0);
        ListNode cur = pre;
        int value;

        while (l1 != null || l2 != null) {
            int a = l1 == null ? 0 : l1.val;
            int b = l2 == null ? 0 : l2.val;
            value = a + b + carry;
            carry = value >= 10 ? 1 : 0;
            value = value % 10;
            cur.next = new ListNode(value);
            cur = cur.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        if (carry == 1) {
            cur.next = new ListNode(1);
        }

        return pre.next;

    }


}

File: python/linked_list/Solution2.py
# 2. 两数相加
from typing import Optional

from linked_list import ListNode


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        res_head = ListNode()
        res = res_head
        while l1 and l2:
            s = l1.val + l2.val + carry
            carry = 0
            if s >= 10:
                s = s % 10
                carry = 1
            res.next = ListNode(s)
            res = res.next
            l1 = l1.next
            l2 = l2.next
        # should consider carry
        while l1:
            s = l1.val + carry
            carry = 0
            if s >= 10:
                s = s % 10
                carry = 1
            res.next = ListNode(s)
            res = res.next
            l1 = l1.next
        while l2:
            s = l2.val + carry
            carry = 0
            if s >= 10:
                s = s % 10
                carry = 1
            res.next = ListNode(s)
            res = res.next
            l2 = l2.next
        if carry == 1:
            res.next = ListNode(1)
        return res_head.next
